File: src/main/java/com/sourcegraph/cody/CodyFileEditorListener.java
package com.sourcegraph.cody;

import com.intellij.openapi.editor.Document;
import com.intellij.openapi.fileEditor.FileDocumentManager;
import com.intellij.openapi.fileEditor.FileEditorManager;
import com.intellij.openapi.fileEditor.FileEditorManagerListener;
import com.intellij.openapi.vfs.VirtualFile;
import com.sourcegraph.cody.agent.CodyAgent;
import com.sourcegraph.cody.agent.CodyAgentClient;
import com.sourcegraph.cody.agent.CodyAgentServer;
import com.sourcegraph.cody.agent.protocol.TextDocument;
import com.sourcegraph.config.ConfigUtil;
import java.util.concurrent.TimeUnit;
import org.jetbrains.annotations.NotNull;

public class CodyFileEditorListener implements FileEditorManagerListener {
  @Override
  public void fileOpened(@NotNull FileEditorManager source, @NotNull VirtualFile file) {
    if (!ConfigUtil.isCodyEnabled()) {
      return;
    }
    Document document = FileDocumentManager.getInstance().getDocument(file);
    if (document == null) {
      return;
    }

    System.err.println("[głupie_logi] public void fileOpened(@NotNull FileEditorManager source, @NotNull VirtualFile file)");
    CodyAgent.getInitializedServer(source.getProject())
        .completeOnTimeout(null, 3, TimeUnit.SECONDS)
        .thenAccept(
            server -> {
                System.err.println("[głupie_logi] fileOpened");
              if (server == null) {
                  System.err.println("[głupie_logi] server == null");
                return;
              }
              if (!CodyAgent.isConnected(source.getProject())) {
                  System.err.println("[głupie_logi] !CodyAgent.isConnected(source.getProject()");
                return;
              }

              server.textDocumentDidOpen(new TextDocument(file.getPath(), document.getText()));

              CodyAgentClient client = CodyAgent.getClient(source.getProject());
              if (client.codebase == null) {
                  System.err.println("[głupie_logi] client.codebase == null");
                return;
              }
              client.codebase.onFileOpened(source.getProject(), file);
                System.err.println("[głupie_logi] client.codebase.onFileOpened(source.getProject(), file);");
            })
            .exceptionally(err -> {
                System.err.println("[głupie_logi] err: " + err);
                   return null;
            });
  }

  @Override
  public void fileClosed(@NotNull FileEditorManager source, @NotNull VirtualFile file) {
    if (!ConfigUtil.isCodyEnabled()) {
      return;
    }
    CodyAgentServer server = CodyAgent.getServer(source.getProject());
    if (server == null) {
      return;
    }
    server.textDocumentDidClose(new TextDocument(file.getPath()));
  }
}

File: src/main/kotlin/com/sourcegraph/cody/listeners/CodyFileEditorListener.kt
package com.sourcegraph.cody.listeners

import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.editor.EditorFactory
import com.intellij.openapi.fileEditor.FileDocumentManager
import com.intellij.openapi.fileEditor.FileEditorManager
import com.intellij.openapi.fileEditor.FileEditorManagerListener
import com.intellij.openapi.fileEditor.TextEditor
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VirtualFile
import com.sourcegraph.cody.agent.CodyAgent
import com.sourcegraph.cody.agent.CodyAgentService.Companion.withAgent
import com.sourcegraph.cody.agent.protocol.ProtocolTextDocument
import com.sourcegraph.cody.agent.protocol.ProtocolTextDocument.Companion.fromVirtualEditorFile

class CodyFileEditorListener : FileEditorManagerListener {
  private val logger = Logger.getInstance(CodyFileEditorListener::class.java)

  override fun fileOpened(source: FileEditorManager, file: VirtualFile) {
    try {
      val textEditor = source.getSelectedEditor(file) as? TextEditor ?: return
      val editor = textEditor.editor
      val protocolTextFile = fromVirtualEditorFile(editor, file)
      EditorChangesBus.documentChanged(editor.project, protocolTextFile)

      withAgent(source.project) { agent: CodyAgent ->
        agent.server.textDocumentDidOpen(protocolTextFile)
      }
    } catch (x: Exception) {
      logger.warn("Error in fileOpened method for file: ${file.path}", x)
    }
  }

  override fun fileClosed(source: FileEditorManager, file: VirtualFile) {
    try {
      val protocolTextFile = ProtocolTextDocument.fromVirtualFile(file)
      EditorChangesBus.documentChanged(source.project, protocolTextFile)
      withAgent(source.project) { agent: CodyAgent ->
        agent.server.textDocumentDidClose(protocolTextFile)
      }
    } catch (x: Exception) {
      logger.warn("Error in fileClosed method for file: ${file.path}", x)
    }
  }

  companion object {
    private val logger = Logger.getInstance(CodyFileEditorListener::class.java)

    // When IDEA starts for the first time, we send duplicate `textDocument/didOpen` notifications
    // with `fileOpened` above. This function is only needed when we restart the agent process.
    fun registerAllOpenedFiles(project: Project, codyAgent: CodyAgent) {

      ApplicationManager.getApplication().invokeLater {
        val fileDocumentManager = FileDocumentManager.getInstance()

        EditorFactory.getInstance().allEditors.forEach { editor ->
          fileDocumentManager.getFile(editor.document)?.let { file ->
            try {
              val textDocument = fromVirtualEditorFile(editor, file)
              codyAgent.server.textDocumentDidOpen(textDocument)
            } catch (x: Exception) {
              logger.warn("Error calling textDocument/didOpen for file: ${file.path}", x)
            }
          }
        }

        if (project.isDisposed) return@invokeLater
        FileEditorManager.getInstance(project).selectedTextEditor?.let { editor ->
          val file = fileDocumentManager.getFile(editor.document)
          try {
            val textDocument = fromVirtualEditorFile(editor, file!!)
            codyAgent.server.textDocumentDidFocus(textDocument)
          } catch (x: Exception) {
            logger.warn("Error calling textDocument/didFocus on ${file?.path}", x)
          }
        }
      }
    }
  }
}
